distribution/tools/geoip-cli/src/test/java/org/elasticsearch/geoip/GeoIpCliTests.java

```diff
@@ -0,0 +1,169 @@
+    }
+
+    public void testNoSource() throws Exception {
+        MockTerminal terminal = new MockTerminal();
+        new GeoIpCli().main(new String[] {}, terminal);
+        assertThat(terminal.getErrorOutput(), containsString("Missing required option(s) [s/source]"));
+    }
+
+    public void testDifferentDirectories() throws Exception {
+        Map<String, byte[]> data = createTestFiles(source);
+
+        GeoIpCli cli = new GeoIpCli();
+        cli.main(new String[] { "-t", target.toAbsolutePath().toString(), "-s", source.toAbsolutePath().toString() }, new MockTerminal());
+
+        try (Stream<Path> list = Files.list(source)) {
+            List<String> files = list.map(p -> p.getFileName().toString()).collect(Collectors.toList());
+            assertThat(files, containsInAnyOrder("a.mmdb", "b.mmdb", "c.tgz"));
+        }
+
+        try (Stream<Path> list = Files.list(target)) {
+            List<String> files = list.map(p -> p.getFileName().toString()).collect(Collectors.toList());
+            assertThat(files, containsInAnyOrder("a.tgz", "b.tgz", "c.tgz", "overview.json"));
+        }
+
+        verifyTarball(data);
+        verifyOverview();
+    }
+
+    public void testSameDirectory() throws Exception {
+        Map<String, byte[]> data = createTestFiles(target);
+
+        GeoIpCli cli = new GeoIpCli();
+        cli.main(new String[] { "-s", target.toAbsolutePath().toString() }, new MockTerminal());
+
+        try (Stream<Path> list = Files.list(target)) {
+            List<String> files = list.map(p -> p.getFileName().toString()).collect(Collectors.toList());
+            assertThat(files, containsInAnyOrder("a.mmdb", "b.mmdb", "a.tgz", "b.tgz", "c.tgz", "overview.json"));
+        }
+
+        Files.delete(target.resolve("a.mmdb"));
+        Files.delete(target.resolve("b.mmdb"));
+
+        verifyTarball(data);
+        verifyOverview();
+    }
+
+    private void verifyOverview() throws Exception {
+        byte[] data = Files.readAllBytes(target.resolve("overview.json"));
+        try (
+            XContentParser parser = XContentType.JSON.xContent()
+                .createParser(NamedXContentRegistry.EMPTY, DeprecationHandler.THROW_UNSUPPORTED_OPERATION, data)
+        ) {
+            @SuppressWarnings({ "unchecked" })
+            List<Map<String, String>> list = (List) parser.list();
+            assertThat(list, containsInAnyOrder(hasEntry("name", "a.tgz"), hasEntry("name", "b.tgz"), hasEntry("name", "c.tgz")));
+            assertThat(list, containsInAnyOrder(hasEntry("url", "a.tgz"), hasEntry("url", "b.tgz"), hasEntry("url", "c.tgz")));
+            for (Map<String, String> map : list) {
+                assertThat(map, hasKey("md5_hash"));
+                assertThat(map, hasKey("updated"));
+            }
+        }
+    }
+
+    private void verifyTarball(Map<String, byte[]> data) throws Exception {
+        // skip tarball verifications on Windows, no tar utility there
+        if (WINDOWS) {
```

Comment:
We could add a tar library as a test dependency and then use that to verify the tarball. That way we don't have to rely on the `tar` command line utility to be available on the machine that runs the tests.

Reply:
Turns out we already have apache common-compress in test framework, I've used it and it makes code much cleaner. Thanks for suggestion!